### ISSUE TITLE
Gems can be hard to find if starting area size is massive

### ISSUE THREAD

**Issue by Deadlock989** [Deadlock989/IndustrialRevolution#93]
**Describe the bug**

Gem-bearing ore patches are always excluded from the "starting area", which it turns out is mainly intended to control how close biter nests can be spawned to the player's starting position. This isn't really clear from the `LuaSurface::get_starting_area_radius()` docs - the main clue is that the "Starting area size" slider is found on the "Enemy" tab of the map settings. This can go all the way up to 600% which it turns out equates to a radius of 900 tiles (over 28 chunks!) 

That said, the autoplacing of ore patches is slightly affected by the starting area (see AutoplaceSpecification). I think the original intention was that gems should never appear in the curated ore patches close to the start position - I assumed that starter ore patches might expand out into a larger starting area but it's not clear that this even happens, and anyhoo, a 28 chunk radius clearly isn't sane. It would be better to ignore the map setting completely and just use a fixed value, e.g. 150 which corresponds to the size of the starting area when the slider is set to the default 100%.

Also, this affects every surface. By default non-Nauvis surfaces don't generate any gems but this can be changed with a remote call already provided (#67) - but there may well be some reason why you want gems everywhere on another non-starting surface.

To do:

- At runtime, store gem exclusion radii per surface in a global table. Also set a default (150). 
- `CONTROL.replace_chunk_ores()` should consult the table and uses the default if there is no set value for a surface ID. A value of 0 would allow gems everywhere, even in the start area.
- Add a new remote call to set the radius per surface.

**Comment by Deadlock989** [comment]
More issues with this, needs rethinking.

- The range should really be a surface specific setting, since there are already remote calls available for other mods to enable or disable gem generation per surface. It may be desirable e.g. to have gems everywhere on an alternative surface.
- In theory other mods can manipulate the start location and the function is currently assuming the vanilla origin of (0,0). Worst case scenario - some sort of multi-force multi-start situation.

**Comment by Deadlock989** [comment]
Multiplayer multi-force competitive/PVP situations are out of scope, too much hassle for something IR2 doesn't cater to. If there are players/teams popping up everywhere in the world, as a compromise the change below will allow third party mods to allow gems everywhere on any given surface.

To do:

- At runtime, store gem exclusion radii per surface in a global table. Also set a default (150). 
- `CONTROL.replace_chunk_ores()` should consult the table and uses the default if there is no set value for a surface ID. A value of 0 would allow gems everywhere, even in the start area.
- Add a new remote call to set the radius per surface.

**Comment by rlundy** [comment]
Interesting.  I had always thought this was an intentional design, that you had to venture into the bug zone to find gems.  Kind of a risk/reward thing: Play it safe with a huge starting area, but then you're going to have to drive far to find gems.  Or play it risky with a small starting area but close gems.

**Comment by Deadlock989** [comment]
I genuinely can't remember exactly why it was done that way - call it post-Covid brain fog or just advancing age. Maybe there was a bit of "let them sally forth" thinking - but I think mostly I was just trying to make gems guaranteed to be unavailable in the starter patches and I misunderstood what that API function related to. 

The problem, if there is one, is that some people can't find gems at all and they don't understand what they're doing wrong, and there's nothing explicit in the map settings about gems to help clue them in, and the in-game doc page about it is a bit gnomic (if people even read it, which they can only do after starting a new game anyhow). I can understand that it must be frustrating that you can't find this one small gate to further progression - and so people just assume the mod is broken.

I guess ideally there would be some sort of explicit slider or toggle in the map settings for gem occurrence (with the status quo as the default) but I don't think the modding API supports that very well. You can add frequency/size/richness or scale/coverage sliders for autoplaced resources (e.g. tin, gold, rubber trees) but as far as I know that's it, you can't add controls of your choice to any particular tab in the settings - I may be wrong. I could change the tooltip for the starting area size slider but again, if people won't read, it won't help them. It could be under Mod Settings > Map but you can't have sliders in those at all and again, people tend not to look at mod settings.

With this fixed radius change there are now three situations:

- Easy - folks make a massive starting area and are much more likely to find gems before they find biters (easier than before).
- Default - folks leave the starting area at default setting (status quo).
- Hard - folks shrink the starting area down to a very challenging size and still have to fight through biters to get to gem-bearing patches (technically harder to get gems than before).

Another weirdness is that if you turn enemies off completely (like I used to do for 3/4 of my Factorio sessions) _after_ you fiddle with the starting area and make it huge, the slider gets grayed out - but the setting is still carried through to the game ... In that situation, as it was before, you'd have a 56 chunk diameter circle of no gems but no biters to fight, which just seems arbitrary.

I have also thought about completely changing the way gems are found to some other scheme but haven't come up with anything I like better.

**Comment by rlundy** [comment]
That all makes sense to me.  Speaking personally, one thing I had trouble with at first in regards to gems was not knowing what to look for.  I think even after I found ore patches with little diamond outlines on the map, it wasn't clear to me that it meant there were gems there.  Hovering over an iron patch shows "1.3M coal"; it doesn't show "1.3M coal + roughly 213 diamonds".  I would imagine that's something that can't be changed.

I wonder if it might help to create a separate Informatron page for gems showing what patches with and without them look like, so at least players know what they're looking for.

Of course, as you say, that only helps if people bother to read it.  😁

If it weren't too much hassle, I wonder if you'd consider a mod option to let the player choose which gem scheme to use?  It could just be a checkbox, where the default setting would be the new fixed-radius setting, and the alternate setting would mean the old version where gems are only found at biter radius.

I was going to ask whether you'd ever considered having gems be an additional byproduct from the ore-washing process, but typing it out, I suddenly got a vague suspicion that that's actually how IR1 worked and I'd just forgotten.

**Comment by Deadlock989** [comment]
> I was going to ask whether you'd ever considered having gems be an additional byproduct from the ore-washing process, but typing it out, I suddenly got a vague suspicion that that's actually how IR1 worked and I'd just forgotten.

You remember correctly. I always felt like having multiple byproducts from washing was very awkward, especially after acid washing became a thing. IR1 games often ended up with having to incinerate or store tens of thousands of unneeded gems. The current "very rare sprinkling + synthesis" scheme was kind of a last resort after all of the following tried and failed:

1. An incredibly overambitious early draft of IR2 had about a hundred kinds of ore patch, ranging from things like pure meteoric iron or copper/tin junkpiles in the starting area that could be smelted or crafted very easily, to oxide ores that need more intensive reduction smelting, to sulphide and carbonate ores that needed multiple stages of processing and waste management, with "nastier" ores rewarded by better yields/density. Not all metals had all the compounds, so that metals had a sort of inherent difficulty/research level - e.g. there was pure copper, copper oxide and copper carbonate, but gold only occurred as a sulphide. Washing for rare metals wasn't a thing - there were lead, nickel and chromium compounds out in the world. Some compounds only occurred in patches of another main metal, e.g. chromium oxide was only ever found in small quantities in specific kinds of iron oxide patch. Crushing for higher yield was not a thing, but some processes had something that was sort of analogous to washing without byproducts. Gems could be found in large quantities throughout very specific patches. Patches also had "quality variants" - low quality patches would churn out more stone and gravel than ore - hence all the variants of patch. There are several reasons this didn't work - arguably it wasn't even fun to make, let alone play - but the relevant one here is that I couldn't get Factorio's default noise/autoplace system to produce anything remotely playable with so many different kinds of patches (massive balancing issues, frequent overlapping patches or total deserts etc.), and I lost patience with trying to write a new noise system or even a slower on_chunk_generated Lua script from scratch.
2. Once I'd settled back on just vanilla + tin + gold, I tried having a duplicate version of all the metal and coal patches which contain tiny quantities of gems throughout. This would have allowed more visual cues on the map and given the kind of estimate of patch content you are talking about above, but it failed for the same main reason as the above - the vanilla ore patch noise function seems to be fine tuned for vanilla's 5 different patches, and can only just cope with IR2's current 7 - pushing it up to 12 produced unplayable or at least unenjoyable maps roughly half the time. I wasted many hours trying to tweak `core/lualib/resource-autoplace.lua` to cope but as a self-taught non-professional programmer ... it wasn't a fun way of spending my time.

> If it weren't too much hassle, I wonder if you'd consider a mod option to let the player choose which gem scheme to use? It could just be a checkbox, where the default setting would be the new fixed-radius setting, and the alternate setting would mean the old version where gems are only found at biter radius.

- [ ] It would not be difficult. I will take a look.

**Comment by Deadlock989** [comment]
OK, I have set up the following:

- By default the new behaviour is to use a gem exclusion radius of 150 (= starting area of 100%) no matter what the starting area slider is set to
- This can be overridden with a checkbox under Map generator > Mod settings > Map settings, which will use whatever value the starting area size is set to, i.e. gems can only be found in biter territory
- Both of these will be overridden on any specific surface if a third party mod uses the new remote call to set it

**Comment by rlundy** [comment]
That sounds great to me.  Well, except the part about diamonds...  What would happen to those massive miners at the end if we had no diamonds to make them with?  Well, presumably you'd keep them but change the recipe.

The gems raining from the sky, though...  How can one keep the game interesting when you've got an impenetrable parameter around your base and the biters are no longer a threat?  Simple: Every so often, there's a ruby/diamond/sapphire storm that randomly smashes bits of your factory.  😂

**Comment by Deadlock989** [comment]
> The gems raining from the sky, though... How can one keep the game interesting when you've got an impenetrable parameter around your base and the biters are no longer a threat? Simple: Every so often, there's a ruby/diamond/sapphire storm that randomly smashes bits of your factory. 😂

Sounds like Space Exploration to me ... 

I have come up with a solution that seems really promising. The gem-bearing ore resource variants have been removed. Instead, copies of the "big rock" entity that randomly litters the world have been created which contain a few gems each. These only spawn on top of specific ore patches in a similar manner to the way the tiny ore patches did, except that this way has the following advantages:

- The gems can be mined from the rocks by hand or robot, instantly. You don't have to plonk a miner down and wait for RNG to grant you a hit on the specific ore resource within its range.
- Gems don't pollute the mining output. No more gems sneaking onto belts and fouling up filters, no more having to put a filter splitter and storage on every outpost.
- The rocks can have a map colour assigned so they can be spotted more easily when exploring the map (no more having to mouse-over every ore patch on the map to spot the little "diamond" shapes, no more wrestling on my part with Blender scripts that try to make sure every ore tile has a visible gem). Technically the old ore variants could have distinct map colours but they were so small the colour had a 50/50 chance of being missed by the checkerboard grid Factorio uses for ore patches.
- All of the existing settings, starting area radius mechanics, existing remote interfaces etc. can be preserved as-is.

I need to play with the numbers a bit to make them rare enough to be interesting but non-frustrating.

**Comment by Deadlock989** [comment]
![sparkles](https://user-images.githubusercontent.com/3128845/191909112-db842318-9ac8-4a77-8364-f04c7e4d5954.gif)

**Comment by rlundy** [comment]
This looks like a great change.  I definitely won't miss finding a stopped production line and discovering that it's because a train waiting on a full load will never get one because there's a single gem in one of its slots.  😄

**Comment by Deadlock989** [comment]
> * By default the new behaviour is to use a gem exclusion radius of 150 (= starting area of 100%) no matter what the starting area slider is set to
> * This can be overridden with a checkbox under Map generator > Mod settings > Map settings, which will use whatever value the starting area size is set to, i.e. gems can only be found in biter territory
> * Both of these will be overridden on any specific surface if a third party mod uses the new remote call to set it

I've changed my mind again and reverted this to the way it was - depends on starting size area only. 

The new method of placing resources is a lot more reliable and the resources are much easier to see on the map, so I don't feel like the fixed radius was really adding anything.

In general I don't like having options that affect gameplay, especially things hidden away in the mod settings - I'm trying to keep IR options purely cosmetic, the one exception being starting tech era.
